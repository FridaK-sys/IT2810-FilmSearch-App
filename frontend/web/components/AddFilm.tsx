import { CreateFormProps } from '../utils/Interface';
import { Center, Button, Modal, Select, CheckIcon } from "native-base";
import { Input, Text } from 'native-base';
import React from 'react';

/** 
* Const for creating a form 
* @param open boolean to decide if modal should be open or closed
* @param onCreate function to create filmitem
* @param onCancel function to close the modal
* @return modal for creating a filmitem
*/
export const CreateForm: React.FC<CreateFormProps> = ({
  open,
  onCreate,
  onCancel,
}) => {
    const [title, setTitle] = React.useState("");
    const [year, setYear] = React.useState("");
    const [cast, setCast] = React.useState("");
    const [genres, setGenres] = React.useState("");

    const handleTitle = (value: string) => setTitle(value);
    const handleYear = (value: string) => setYear(value);
    const handleCast = (value: string) => setCast(value);
    const handleGenres = (value: string) => setGenres(value);

    function createFilm() {
      onCreate({
          title: title, 
          year: year, 
          cast: cast,
          genres: genres,
        }
      )
    }

    return(
        <Center>
        <Modal isOpen={open} onClose={onCancel}>
        <Modal.Content maxWidth="400px">
          <Modal.CloseButton />
          <Modal.Header><Text>Chosen film</Text></Modal.Header>
          <Modal.Body>
            <Input
                marginBottom={2}  
                variant="outline"
                onChangeText={handleTitle}
                placeholder="Title"
            />
            <Input
                marginBottom={2}
                variant="outline"
                onChangeText={handleYear}
                placeholder="Year"
            />
            <Input
                marginBottom={2}
                variant="outline"
                onChangeText={handleCast}
                placeholder="Cast"
            />
            <Select 
              placeholder="Genres" 
              mx={{base: 0, md: "Genre"}} 
              onValueChange={handleGenres} 
              _selectedItem={{
                bg: "cyan.600",
                  endIcon: <CheckIcon size={4} />
              }} 
              accessibilityLabel="Select genre">
                <Select.Item label="Drama" value="Drama" />
                <Select.Item label="Documentary" value="Documentary" />
                <Select.Item label="Sports" value="Sports" />
                <Select.Item label="Silent" value="Silent" />
                <Select.Item label="Adventure" value="Adventure" />
                <Select.Item label="Western" value="Western" />
                <Select.Item label="Romance" value="Romance" />
                <Select.Item label="War" value="War" />
                <Select.Item label="Comedy" value="Comedy" />
                <Select.Item label="Horror" value="Horror" />
                <Select.Item label="Historical" value="Historical" />
                <Select.Item label="Animated" value="Animated" />
            </Select>
          </Modal.Body>
          <Modal.Footer>
            <Button marginRight={2} variant="outline" colorScheme="blueGray" onPress={onCancel}>
                <Text>Close</Text>
            </Button>
            <Button variant="outline" colorScheme="blueGray" onPress={createFilm}>
                <Text>Add film</Text>
            </Button>
          </Modal.Footer>
        </Modal.Content>
      </Modal>
    </Center>
    );
};