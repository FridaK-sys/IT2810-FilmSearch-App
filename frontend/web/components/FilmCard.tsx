import React from "react";
import { CreateFilmCard } from "../utils/Interface";
import { Box, Text, VStack, Pressable, Divider } from 'native-base';

/** 
* Const for creating a filmcard 
* @param film the film to be displayed on the card
* @param handleClick function to handle clicking on a film
* @return a filmcard
*/
export const FilmCard: React.FC<CreateFilmCard> = ({
    film,
    handleClick
}) => {
    return(
        <Box borderColor={"600"} backgroundColor="white" borderWidth="1" borderRadius="md" width="85%" margin={5} marginBottom={1}>   
            <Pressable onPress={() => handleClick(film)}>
                <VStack divider={<Divider />} width="100%">
                    <Box padding={3} px="4" pt="4">
                        <Text fontSize={20}>{film.title}</Text>
                    </Box>
                    <Box px="4" padding={6}>
                        <Text fontSize={15}>Year Released: {film.year? film.year: "No year is registered"}</Text>
                    </Box>
                </VStack>
            </Pressable>
        </Box>
    );
}