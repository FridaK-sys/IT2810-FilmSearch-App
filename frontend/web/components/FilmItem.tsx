import { Button, Center, Modal, Text } from "native-base";
import { CreateShowFilmProps } from "../utils/Interface";
import React from "react";

/** 
* Const for showing a modal with more information on a filmitem
* @param film chosen filmitem
* @param open boolean to decide if modal should be open or closed
* @param onCancel function to close the modal
* @return a modal to show information about a filmitem 
*/
export const ShowFilmItem: React.FC<CreateShowFilmProps> = ({
    film,
    open,
    onCancel,
  }) => {
    return(
        <Center>
            <Modal isOpen={open} onClose={onCancel}>
                <Modal.Content maxWidth="400px">
                    <Modal.CloseButton />
                    <Modal.Header>Chosen film</Modal.Header>
                    <Modal.Body>
                            <Text paddingTop={1} marginBottom={2} fontSize={22}>{film.title? film.title: "No title is registered for this film"} </Text>
                            <Text marginBottom={2} fontSize={16}>Year: {film.year? film.year: "No year is registered for this film"} </Text>
                            <Text marginBottom={2} fontSize={16}>Cast: {(film.cast.length === 0)? "No cast is registered for this film": film.cast.map((el) => el + ", ")}</Text>
                            <Text marginBottom={2} fontSize={16}>Genres: {(film.genres.length === 0)? "No genre is registered for this film" : film.genres.map((el) => el + ", ")}</Text>
                    </Modal.Body>
                    <Modal.Footer>
                        <Button borderColor={"600"} borderWidth="1" variant="outline" colorScheme="blueGray" onPress={onCancel}>
                            <Text fontSize={14}>Close</Text>
                        </Button>
                    </Modal.Footer>
                </Modal.Content>
            </Modal>    
        </Center>
    );
};