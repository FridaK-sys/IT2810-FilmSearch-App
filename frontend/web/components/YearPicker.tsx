import { Button, Center, Modal, Text } from "native-base";

import { CreateShowYearProps } from "../utils/Interface";
import { Picker } from 'react-native-wheel-pick';
import React from "react";
import { setYear } from '../redux/actions';
import { useDispatch } from "react-redux";

/** 
* Const for showing a wheel picker for year filtering
* @param open boolean to decide if modal should be open or closed
* @param onClose function to close the modal
* @return a modal to choose year filter
*/
export const YearPicker: React.FC<CreateShowYearProps> = ({
    open,
    onClose,
  }) => {
    const [tempYear, setTempYear] = React.useState<string>('0');
    const dispatch = useDispatch();
    const yearList : string[] = [];

    for (let i = 1900; i < 2019; i++) {
    yearList.push(i.toString());
    }

    const pickYear = () => {
        dispatch(setYear(tempYear));
        onClose();
      };
  
    return (
        <Center>
            <Modal isOpen={open} onClose={onClose}>
            <Modal.Content maxWidth="400px">
            <Picker
                            style={{ backgroundColor: 'white', width: 300, height: 215 }}
                            selectedValue='0'
                            pickerData={yearList}
                            onValueChange={(e: string) => setTempYear(e)}
                        />
            <Modal.Footer>
            <Button variant="ghost" colorScheme="blueGray" onPress={() => pickYear()}>
                <Text>Choose year</Text>
            </Button>
          </Modal.Footer>
            </Modal.Content>
        </Modal>
        </Center>
    );
}