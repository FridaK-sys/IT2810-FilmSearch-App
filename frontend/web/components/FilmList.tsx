import { ADD_FILM, SEARCH_FILMS } from '../queries/filmQueries';
import { Box, Button, CheckIcon, Divider, HStack, Heading, Input, InputRightElement, Row, Select, Spinner, Text, VStack } from 'native-base';
import { Film, Values } from '../utils/Interface';
import { setGenre, setSorting, setTitle, setYear } from '../redux/actions';
import { useDispatch, useSelector } from "react-redux";
import { useMutation, useQuery } from '@apollo/client'

import { CreateForm } from './AddFilm';
import { Pressable } from "react-native";
import React from 'react';
import { ShowFilmItem } from './FilmItem';
import {Store} from "../redux/store";
import {YearPicker} from './YearPicker';
import { useState } from 'react';

const PAGE_SIZE = 15;

/** 
* Main component to show list of filmitems on React application
*/
export default function FilmList() {
    const [page, setPage] = useState(0);
    const [openCreate, setOpenCreate] = useState(false);
    const [currentPost, setCurrentPost] = useState<Film>({
        _id: "",
        title: "",
        year: "",
        cast: [""],
        genres: [""]
    });
    const [showFilm, setShowFilm] = useState(false);
    const [filterYear, setFilterYear] = useState(false);
    const [searchTitle, setSearchTitle] = useState ("");

    const hideFilm = () => {
      setShowFilm(false);
    };

    const dispatch = useDispatch();
    const [createPost] = useMutation(ADD_FILM);

    let title = useSelector ((state: Store) => state.title); //fetching title filter from redux store
    let genre = useSelector ((state: Store) => state.genre); //fetching genre filter from redux store
    let year = useSelector ((state: Store) => state.year); //fetching year filter from redux store
    let sorting = useSelector ((state: Store) => state.sorting); //fetching sorting filter from redux store
    
    /** 
    * Retrieves data, loading and error from graphql server
    * @param variables to be considered when retreiving data
    * @return data from graphql server 
    */
    const { loading, error, data } = useQuery(SEARCH_FILMS, {
        variables: {
            limit: PAGE_SIZE,
            offset: page * PAGE_SIZE,
            titleFilter: title,
            genreFilter: genre,
            yearFilter: parseInt(year, 10),
            sorting: parseInt(sorting, 10),
        },
    });

    if (loading) {
        return (
            <HStack space={2} justifyContent="center">
                <Spinner accessibilityLabel="Loading posts" />
                    <Heading color="primary.500" fontSize="md">
                        Loading
                    </Heading>
            </HStack>
        )
    }

   if (error) {
        console.log(error)
        return (
            <h5 style={{color: "#ffffff"}}>Something went wrong when trying to connect to the server...</h5>
        )
    }

    /** 
    * Creates a filmitem in the database
    * @param film to be created
    */
    const onCreate = (film: Values) => { 
        createPost({
            variables: {
                title: film.title,
                year: film.year? parseInt(film.year, 10) : 0,
                cast: film.cast? film.cast.split(",") : [],
                genres: film.genres? [film.genres]: [],
            }
        });
        dispatch(setTitle(film.title))
        setOpenCreate(false);
    };

    function handleClick(post: Film) {
        setCurrentPost(post);
        setShowFilm(true);
    }

    let body: any = [];
    body.push (
        data.getFilteredPosts?.map((post: Film) => (
            <Box borderRadius="md">
                <Pressable onPress={() => handleClick(post)}>
            <VStack space="4" divider={<Divider />}>
                <Box px="4" marginBottom={10} marginTop={10}>
                    <Heading marginBottom={2}>
                        <Text>{post.title}</Text>
                    </Heading>
                    <Text>Year Released: {post.year? post.year: ""}</Text>
                </Box>
            </VStack>
            </Pressable>
            </Box>
        ))
    )

    /** 
    * Resets the filters in redux
    */
    function useReset() {
        dispatch(setTitle(""))
        dispatch(setGenre(""))
        dispatch(setYear("0"))
        dispatch(setSorting("1"))

        setSearchTitle("");
        title = "";
        genre = "";
        year = "0";
        sorting = "1";
    }
    
    return (
        <>
        {!loading && !error && 
            <Box justifyContent="center">    
                <Box justifyContent="center">
                    <Box marginTop={75} margin={2}>
                        <Input 
                            value={searchTitle}
                            placeholder="Search for title" 
                            onChangeText={(e: string) => setSearchTitle(e)}
                            InputRightElement={<Button size="xs" rounded="none" w="1/5" h="full" onPress={() => dispatch(setTitle(searchTitle))}>
                            {<Text>Search</Text>}
                            </Button>}
                        />
                    </Box>
                    <Box margin={2}>
                        <Select selectedValue={genre} mx={{base: 0, md: "Genre"
                            }} placeholder="Filter by genre"  onValueChange={e => dispatch(setGenre(e))} _selectedItem={{
                        bg: "cyan.600",
                            endIcon: <CheckIcon size={4} />
                        }} accessibilityLabel="Select genre">
                            <Select.Item label="Drama" value="Drama" />
                            <Select.Item label="Documentary" value="Documentary" />
                            <Select.Item label="Sports" value="Sports" />
                            <Select.Item label="Silent" value="Silent" />
                            <Select.Item label="Adventure" value="Adventure" />
                            <Select.Item label="Western" value="Western" />
                            <Select.Item label="Romance" value="Romance" />
                            <Select.Item label="War" value="War" />
                            <Select.Item label="Comedy" value="Comedy" />
                            <Select.Item label="Horror" value="Horror" />
                            <Select.Item label="Historical" value="Historical" />
                            <Select.Item label="Animated" value="Animated" />
                        </Select>
                    </Box>
                    <Box margin={2}>
                        <Select selectedValue={sorting} mx={{base: 0, md: "Sort"
                            }} onValueChange={e => dispatch(setSorting(e))} _selectedItem={{
                        bg: "cyan.600",
                            endIcon: <CheckIcon size={4} />
                            }} accessibilityLabel="Sort">
                            <Select.Item label="Ascending" value="1" />
                            <Select.Item label="Descending" value="-1" />
                        </Select>
                    </Box>
                    <Box margin={2}>
                        <Button   
                            onPress={() => {
                                setFilterYear(true);
                            }}
                        >
                            <Text>Filter on year</Text>
                        </Button>
                    </Box>
                    <YearPicker
                            open={filterYear}
                            onClose={() => setFilterYear(false)}
                    />
                    <Box margin={2}>
                        <Button 
                            onPress={useReset}
                        >
                            <Text>Reset Filters</Text> 
                        </Button>
                    </Box>
                    <Box margin={2}>
                        <Button
                            onPress={() => {
                                setOpenCreate(true);
                            }}
                        >
                            <Text>Add New Film</Text>
                        </Button>
                        <CreateForm
                            open={openCreate}
                            onCreate={onCreate}
                            onCancel={() => {
                                setOpenCreate(false);
                            }}
                        /> 
                    </Box>
                </Box>

                <Row justifyContent="center" alignItems = 'center'>
                    {body}
                </Row>

                <ShowFilmItem 
                    film={currentPost} 
                    open={showFilm} 
                    onCancel={hideFilm} 
                />

                <Button
                    margin={2}
                    disabled={loading}
                    onPress={() => (setPage(prev => prev-1))}
                >
                    <Text>Previous</Text>
                </Button>

                <Button
                    margin={2}
                    disabled={loading}
                    onPress={() => (setPage(prev => prev+1))}
                >
                    <Text>Next</Text>
                </Button>
            </Box>
        }
        </>
    )
};
