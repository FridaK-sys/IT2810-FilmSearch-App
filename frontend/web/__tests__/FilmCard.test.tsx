import '@testing-library/jest-dom'
import { render, screen, } from '@testing-library/react';
import {FilmCard} from '../components/FilmCard';
import { NativeBaseProvider } from 'native-base';
import { Provider } from 'react-redux';
import React from 'react';
import renderer from "react-test-renderer";
import store from '../redux/store';

const film = {"_id":"63453c374d8655244d2aa472","title":"After Dark in Central Park",
"year":"1900","cast":[],"genres":[]}
const handleClick = jest.fn();

const inset = {
    frame: { x: 0, y: 0, width: 0, height: 0 },
    insets: { top: 0, left: 0, right: 0, bottom: 0 },
  };
/** 
* Test for loading card and that it shows correct information
*/
it("shows information about film" , async () => {
    render(
    <NativeBaseProvider initialWindowMetrics={inset}>
    <Provider store={store}>
      <FilmCard film={film} handleClick={handleClick}/>
    </Provider>
    </NativeBaseProvider>
  );
  expect(await screen.getByTestId("title")).toHaveTextContent("After Dark in Central Park");
  expect(await screen.getByTestId("year")).toHaveTextContent("Year Released: 1900");
});

/** 
* Test for that it opens a filmitem when clicked on
*/
it("runs handleClick function when clicked on" , async () => {
  const card = render(
  <NativeBaseProvider initialWindowMetrics={inset}>
  <Provider store={store}>
    <FilmCard film={film} handleClick={handleClick}/>
  </Provider>
  </NativeBaseProvider>
);
  card.getByText("After Dark in Central Park").click();
  expect(handleClick.mock.calls.length).toBe(1);
});

/** 
* Snapshot test for loading card
*/
describe("Jest Snapshot testing suite",  () => {
    it("Matches DOM Snapshot",   () => {
      const domTree = renderer.create(
      <NativeBaseProvider initialWindowMetrics={inset}><Provider store={store}>
      <FilmCard film={film} handleClick={handleClick}/></Provider></NativeBaseProvider>).toJSON();
      expect(domTree).toMatchSnapshot();
  
      
    });
  });

