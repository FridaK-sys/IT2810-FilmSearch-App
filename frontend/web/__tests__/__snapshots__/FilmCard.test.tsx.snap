// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot testing suite Matches DOM Snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      {
        "borderColor": "#e5e7eb",
        "borderRadius": 6,
        "borderWidth": 1,
        "marginBottom": 4,
        "marginLeft": 20,
        "marginRight": 20,
        "marginTop": 20,
        "width": "85%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      dataSet={{}}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={{}}
    >
      <View
        dataSet={{}}
        style={
          {
            "width": "100%",
          }
        }
      >
        <View
          dataSet={{}}
          style={
            {
              "paddingBottom": 12,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
            }
          }
        >
          <Text
            data-testid="title"
            dataSet={{}}
            style={
              {
                "backgroundColor": undefined,
                "color": "#171717",
                "fontFamily": undefined,
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 21,
                "textDecorationLine": undefined,
              }
            }
          >
            After Dark in Central Park
          </Text>
        </View>
        <View
          aria-orientation="horizontal"
          dataSet={{}}
          style={
            {
              "backgroundColor": "#d4d4d4",
              "height": 1,
              "width": "100%",
            }
          }
          thickness="1"
        />
        <View
          dataSet={{}}
          style={
            {
              "paddingBottom": 24,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 24,
            }
          }
        >
          <Text
            data-testid="year"
            dataSet={{}}
            style={
              {
                "backgroundColor": undefined,
                "color": "#171717",
                "fontFamily": undefined,
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 21,
                "textDecorationLine": undefined,
              }
            }
          >
            Year Released: 
            1900
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
